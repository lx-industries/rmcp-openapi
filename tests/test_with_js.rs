use rmcp::transport::streamable_http_server::session::local::LocalSessionManager;
use rmcp::transport::{SseServer, StreamableHttpServerConfig, StreamableHttpService};
use rmcp_openapi::OpenApiServer;
use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
use url::Url;

mod common;
use common::mock_server::MockPetstoreServer;
use mockito::Mock;
use serde_json::json;

/// Create a petstore server with base URL for HTTP requests
fn create_petstore_mcp_server_with_base_url(base_url: Url) -> anyhow::Result<OpenApiServer> {
    let spec_content = include_str!("assets/petstore-openapi.json");
    let spec_url = Url::parse("test://petstore")?;
    let mut server =
        OpenApiServer::with_base_url(rmcp_openapi::OpenApiSpecLocation::Url(spec_url), base_url)?;

    // Parse the embedded spec
    let json_value: serde_json::Value = serde_json::from_str(spec_content).unwrap();
    let spec = rmcp_openapi::openapi::OpenApiSpec::from_value(json_value).unwrap();
    server.register_spec(spec).unwrap();

    Ok(server)
}

const SSE_BIND_ADDRESS: &str = "127.0.0.1:8000";
const STREAMABLE_HTTP_BIND_ADDRESS: &str = "127.0.0.1:8001";

async fn init() -> anyhow::Result<()> {
    let _ = tracing_subscriber::registry()
        .with(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| "debug".to_string().into()),
        )
        .with(tracing_subscriber::fmt::layer())
        .try_init();
    tokio::process::Command::new("npm")
        .arg("install")
        .current_dir("tests/test_with_js")
        .spawn()?
        .wait()
        .await?;
    Ok(())
}

#[tokio::test]
async fn test_with_js_sse_client() -> anyhow::Result<()> {
    init().await?;

    // Start mock server for HTTP requests
    let mut mock_server = MockPetstoreServer::new_with_port(8084).await;

    // Set up mock responses for all tool calls
    let _get_pet_mock = mock_server.mock_get_pet_by_id(123);
    let _find_pets_mock = mock_server.mock_find_pets_by_multiple_status();
    let _add_pet_mock = mock_server.mock_add_pet();
    let _error_mock = mock_server.mock_get_pet_by_id_not_found(999999);
    let _validation_error_mock = mock_server.mock_add_pet_validation_error();

    // Start MCP server with mock API base URL
    let base_url = mock_server.base_url();
    let ct = SseServer::serve(SSE_BIND_ADDRESS.parse()?)
        .await?
        .with_service(move || create_petstore_mcp_server_with_base_url(base_url.clone()).unwrap());

    let output = tokio::process::Command::new("node")
        .arg("client.js")
        .current_dir("tests/test_with_js")
        .output()
        .await?;

    if !output.status.success() {
        eprintln!("JavaScript client failed:");
        eprintln!("STDOUT: {}", String::from_utf8_lossy(&output.stdout));
        eprintln!("STDERR: {}", String::from_utf8_lossy(&output.stderr));
    }
    assert!(output.status.success());

    // Capture and validate the actual MCP responses
    let stdout = String::from_utf8(output.stdout)?;
    let mut responses: Vec<serde_json::Value> = stdout
        .lines()
        .filter(|line| !line.is_empty())
        .map(serde_json::from_str)
        .collect::<Result<Vec<_>, _>>()?;

    // Sort arrays for deterministic snapshots (preserve_order handles object properties)
    for response in &mut responses {
        if let Some(tools) = response
            .get_mut("data")
            .and_then(|d| d.get_mut("tools"))
            .and_then(|t| t.as_array_mut())
        {
            tools.sort_by(|a, b| {
                let name_a = a.get("name").and_then(|n| n.as_str()).unwrap_or("");
                let name_b = b.get("name").and_then(|n| n.as_str()).unwrap_or("");
                name_a.cmp(name_b)
            });
        }
    }

    insta::assert_json_snapshot!("js_sse_client_responses", responses);
    ct.cancel();
    Ok(())
}

#[tokio::test]
async fn test_with_js_streamable_http_client() -> anyhow::Result<()> {
    init().await?;

    // Start mock server for HTTP requests - use unique port for StreamableHTTP test
    let mut mock_server = MockPetstoreServer::new_with_port(8085).await;

    // Set up mock responses for all tool calls
    let _get_pet_mock = mock_server.mock_get_pet_by_id(123);
    let _find_pets_mock = mock_server.mock_find_pets_by_multiple_status();
    let _add_pet_mock = mock_server.mock_add_pet();
    let _error_mock = mock_server.mock_get_pet_by_id_not_found(999999);
    let _validation_error_mock = mock_server.mock_add_pet_validation_error();

    let base_url = mock_server.base_url();
    let service = StreamableHttpService::new(
        move || Ok(create_petstore_mcp_server_with_base_url(base_url.clone()).unwrap()),
        std::sync::Arc::new(LocalSessionManager::default()),
        StreamableHttpServerConfig {
            stateful_mode: true,
            sse_keep_alive: None,
        },
    );

    let router = axum::Router::new().nest_service("/mcp", service);
    let tcp_listener = tokio::net::TcpListener::bind(STREAMABLE_HTTP_BIND_ADDRESS).await?;
    let ct = tokio_util::sync::CancellationToken::new();

    let server_handle = tokio::spawn({
        let ct = ct.clone();
        async move {
            let _ = axum::serve(tcp_listener, router)
                .with_graceful_shutdown(async move { ct.cancelled_owned().await })
                .await;
        }
    });

    // Give the server a moment to start
    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;

    let output = tokio::process::Command::new("node")
        .arg("streamable_client.js")
        .current_dir("tests/test_with_js")
        .output()
        .await?;

    if !output.status.success() {
        eprintln!("StreamableHttp client failed:");
        eprintln!("stdout: {}", String::from_utf8_lossy(&output.stdout));
        eprintln!("stderr: {}", String::from_utf8_lossy(&output.stderr));
    }
    assert!(output.status.success());

    // Capture and validate the actual MCP responses
    let stdout = String::from_utf8(output.stdout)?;
    let mut responses: Vec<serde_json::Value> = stdout
        .lines()
        .filter(|line| !line.is_empty())
        .map(serde_json::from_str)
        .collect::<Result<Vec<_>, _>>()?;

    // Sort arrays for deterministic snapshots (preserve_order handles object properties)
    for response in &mut responses {
        if let Some(tools) = response
            .get_mut("data")
            .and_then(|d| d.get_mut("tools"))
            .and_then(|t| t.as_array_mut())
        {
            tools.sort_by(|a, b| {
                let name_a = a.get("name").and_then(|n| n.as_str()).unwrap_or("");
                let name_b = b.get("name").and_then(|n| n.as_str()).unwrap_or("");
                name_a.cmp(name_b)
            });
        }
    }

    insta::assert_json_snapshot!("js_streamable_http_client_responses", responses);

    ct.cancel();
    server_handle.await?;
    Ok(())
}

// Test-specific mock methods for MockPetstoreServer
impl MockPetstoreServer {
    /// Mock successful getPetById response
    pub fn mock_get_pet_by_id(&mut self, pet_id: u64) -> Mock {
        let pet_response = json!({
            "id": pet_id,
            "name": "doggie",
            "category": {
                "id": 1,
                "name": "Dogs"
            },
            "photoUrls": ["https://example.com/photo1.jpg"],
            "tags": [
                {
                    "id": 1,
                    "name": "tag1"
                }
            ],
            "status": "available"
        });

        self.server
            .mock("GET", format!("/pet/{pet_id}").as_str())
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(pet_response.to_string())
            .create()
    }

    /// Mock getPetById with 404 Not Found
    pub fn mock_get_pet_by_id_not_found(&mut self, pet_id: u64) -> Mock {
        self.server
            .mock("GET", format!("/pet/{pet_id}").as_str())
            .with_status(404)
            .with_header("content-type", "application/json")
            .with_body(json!({"message": "Pet not found"}).to_string())
            .create()
    }

    /// Mock findPetsByStatus response for multiple status values
    pub fn mock_find_pets_by_multiple_status(&mut self) -> Mock {
        let pets_response = json!([
            {
                "id": 1,
                "name": "doggie",
                "category": {
                    "id": 1,
                    "name": "Dogs"
                },
                "photoUrls": ["https://example.com/photo1.jpg"],
                "tags": [
                    {
                        "id": 1,
                        "name": "tag1"
                    }
                ],
                "status": "available"
            },
            {
                "id": 2,
                "name": "kitty",
                "category": {
                    "id": 2,
                    "name": "Cats"
                },
                "photoUrls": ["https://example.com/photo2.jpg"],
                "tags": [
                    {
                        "id": 2,
                        "name": "tag2"
                    }
                ],
                "status": "pending"
            }
        ]);

        self.server
            .mock("GET", "/pet/findByStatus")
            .match_query(mockito::Matcher::AnyOf(vec![
                // Match multiple status values in query
                mockito::Matcher::Regex(r"status=available.*status=pending".to_string()),
                mockito::Matcher::Regex(r"status=pending.*status=available".to_string()),
                // Fallback to any status query
                mockito::Matcher::Regex(r"status=.+".to_string()),
            ]))
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(pets_response.to_string())
            .create()
    }

    /// Mock successful addPet response
    pub fn mock_add_pet(&mut self) -> Mock {
        let pet_response = json!({
            "id": 123,
            "name": "new doggie",
            "category": {
                "id": 1,
                "name": "Dogs"
            },
            "photoUrls": ["https://example.com/photo1.jpg"],
            "tags": [
                {
                    "id": 1,
                    "name": "tag1"
                }
            ],
            "status": "available"
        });

        self.server
            .mock("POST", "/pet")
            .match_header("content-type", "application/json")
            .with_status(201)
            .with_header("content-type", "application/json")
            .with_body(pet_response.to_string())
            .create()
    }

    /// Mock addPet with validation error
    pub fn mock_add_pet_validation_error(&mut self) -> Mock {
        self.server
            .mock("POST", "/pet")
            .with_status(400)
            .with_header("content-type", "application/json")
            .with_body(
                json!({
                    "message": "Invalid input",
                    "details": "Name is required"
                })
                .to_string(),
            )
            .create()
    }
}
