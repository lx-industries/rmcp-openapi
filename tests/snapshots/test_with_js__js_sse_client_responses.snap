---
source: tests/test_with_js.rs
expression: responses
---
[
  {
    "data": {
      "tools": [
        {
          "description": "Add a new pet to the store\n\nAdd a new pet to the store\n\nEndpoint: POST /pet",
          "inputSchema": {
            "additionalProperties": false,
            "properties": {
              "request_body": {
                "additionalProperties": true,
                "description": "Request body data (JSON)",
                "type": "object",
                "x-content-type": "application/json",
                "x-location": "body"
              }
            },
            "required": [],
            "type": "object"
          },
          "name": "addPet"
        },
        {
          "description": "Finds Pets by status\n\nMultiple status values can be provided with comma separated strings\n\nEndpoint: GET /pet/findByStatus",
          "inputSchema": {
            "additionalProperties": false,
            "properties": {
              "status": {
                "description": "Status values that need to be considered for filter",
                "items": {
                  "type": "string"
                },
                "type": "array",
                "x-parameter-location": "query",
                "x-parameter-required": true
              },
              "timeout_seconds": {
                "default": 30,
                "description": "Request timeout in seconds",
                "maximum": 300,
                "minimum": 1,
                "type": "integer"
              }
            },
            "required": [
              "status"
            ],
            "type": "object"
          },
          "name": "findPetsByStatus"
        },
        {
          "description": "Find pet by ID\n\nReturns a single pet\n\nEndpoint: GET /pet/{petId}",
          "inputSchema": {
            "additionalProperties": false,
            "properties": {
              "petId": {
                "description": "ID of pet to return",
                "format": "int64",
                "minimum": 1,
                "type": "integer",
                "x-parameter-location": "path",
                "x-parameter-required": true
              }
            },
            "required": [
              "petId"
            ],
            "type": "object"
          },
          "name": "getPetById"
        }
      ]
    },
    "type": "tools_list"
  },
  {
    "data": {
      "resources": []
    },
    "type": "resources_list"
  },
  {
    "data": {
      "resourceTemplates": []
    },
    "type": "resource_templates_list"
  },
  {
    "data": {
      "prompts": []
    },
    "type": "prompts_list"
  },
  {
    "arguments": {
      "petId": 123
    },
    "data": {
      "content": [
        {
          "text": "HTTP ✅ Success\n\nStatus: 200 OK\n\nRequest: GET http://127.0.0.1:8084/pet/123\n\nResponse Body:\n{\n  \"category\": {\n    \"id\": 1,\n    \"name\": \"Dogs\"\n  },\n  \"id\": 123,\n  \"name\": \"doggie\",\n  \"photoUrls\": [\n    \"https://example.com/photo1.jpg\"\n  ],\n  \"status\": \"available\",\n  \"tags\": [\n    {\n      \"id\": 1,\n      \"name\": \"tag1\"\n    }\n  ]\n}",
          "type": "text"
        }
      ],
      "isError": false
    },
    "success": true,
    "tool": "getPetById",
    "type": "tool_call_result"
  },
  {
    "arguments": {
      "status": [
        "available",
        "pending"
      ]
    },
    "data": {
      "content": [
        {
          "text": "HTTP ✅ Success\n\nStatus: 200 OK\n\nRequest: GET http://127.0.0.1:8084/pet/findByStatus?status=available&status=pending\n\nResponse Body:\n[\n  {\n    \"category\": {\n      \"id\": 1,\n      \"name\": \"Dogs\"\n    },\n    \"id\": 1,\n    \"name\": \"doggie\",\n    \"photoUrls\": [\n      \"https://example.com/photo1.jpg\"\n    ],\n    \"status\": \"available\",\n    \"tags\": [\n      {\n        \"id\": 1,\n        \"name\": \"tag1\"\n      }\n    ]\n  },\n  {\n    \"category\": {\n      \"id\": 2,\n      \"name\": \"Cats\"\n    },\n    \"id\": 2,\n    \"name\": \"kitty\",\n    \"photoUrls\": [\n      \"https://example.com/photo2.jpg\"\n    ],\n    \"status\": \"pending\",\n    \"tags\": [\n      {\n        \"id\": 2,\n        \"name\": \"tag2\"\n      }\n    ]\n  }\n]",
          "type": "text"
        }
      ],
      "isError": false
    },
    "success": true,
    "tool": "findPetsByStatus",
    "type": "tool_call_result"
  },
  {
    "arguments": {
      "request_body": {
        "name": "MCP Test Dog",
        "status": "available"
      }
    },
    "data": {
      "content": [
        {
          "text": "HTTP ✅ Success\n\nStatus: 201 Created\n\nRequest: POST http://127.0.0.1:8084/pet\n\nRequest Body:\n{\n  \"category\": {\n    \"id\": 1,\n    \"name\": \"Dogs\"\n  },\n  \"name\": \"MCP Test Dog\",\n  \"photoUrls\": [\n    \"https://example.com/mcp-test-dog.jpg\"\n  ],\n  \"status\": \"available\",\n  \"tags\": [\n    {\n      \"id\": 1,\n      \"name\": \"mcp-test\"\n    }\n  ]\n}\n\nResponse Body:\n{\n  \"category\": {\n    \"id\": 1,\n    \"name\": \"Dogs\"\n  },\n  \"id\": 123,\n  \"name\": \"new doggie\",\n  \"photoUrls\": [\n    \"https://example.com/photo1.jpg\"\n  ],\n  \"status\": \"available\",\n  \"tags\": [\n    {\n      \"id\": 1,\n      \"name\": \"tag1\"\n    }\n  ]\n}",
          "type": "text"
        }
      ],
      "isError": false
    },
    "success": true,
    "tool": "addPet",
    "type": "tool_call_result"
  },
  {
    "arguments": {
      "petId": 999999
    },
    "data": {
      "content": [
        {
          "text": "HTTP ❌ Error\n\nStatus: 404 Not Found - Not Found: Endpoint or resource does not exist\n\nRequest: GET http://127.0.0.1:8084/pet/999999\n\nResponse Body:\n{\n  \"message\": \"Pet not found\"\n}",
          "type": "text"
        }
      ],
      "isError": true
    },
    "success": true,
    "tool": "getPetById",
    "type": "tool_call_result"
  },
  {
    "arguments": {
      "request_body": {
        "status": "invalid_status_value"
      }
    },
    "data": {
      "content": [
        {
          "text": "HTTP ❌ Error\n\nStatus: 400 Bad Request - Bad Request: Check request parameters\n\nRequest: POST http://127.0.0.1:8084/pet\n\nRequest Body:\n{\n  \"status\": \"invalid_status_value\"\n}\n\nResponse Body:\n{\n  \"details\": \"Name is required\",\n  \"message\": \"Invalid input\"\n}",
          "type": "text"
        }
      ],
      "isError": true
    },
    "success": true,
    "tool": "addPet",
    "type": "tool_call_result"
  }
]
